dnsext-types/DNS/Types/Domain.hs

```diff
@@ -39,226 +35,225 @@ import DNS.Types.Imports
+    rep = foldr (\l r -> escapeLabel _period l <> "." <> r) "" ls
+
 instance Eq Domain where
-    d0 == d1 = lowerDomain d0 == lowerDomain d1
+    d0 == d1 = canonicalLabels d0 == canonicalLabels d1
 
 instance Ord Domain where
     d0 <= d1 = canonicalLabels d0 <= canonicalLabels d1
 
 instance Show Domain where
-    show d = "\"" ++ origName d ++ "\""
+    show d = "\"" ++ toRepresentation d ++ "\""
 
 instance IsString Domain where
-    fromString = ciName
+    fromString = fromRepresentation
 
 instance Semigroup Domain where
-    d0 <> d1 = domain (origDomain d0 <> origDomain d1)
+    d0 <> d1 = domainFromWireLabels (wireLabels d0 <> wireLabels d1)
 
-instance CaseInsensitiveName Domain ShortByteString where
-    ciName o = domain o
-    origName  d = origDomain d
-    lowerName d = lowerDomain d
+instance IsRepresentation Domain ShortByteString where
+    fromRepresentation o = domain o
+    toRepresentation  d = representation d
 
-instance CaseInsensitiveName Domain ByteString where
-    ciName o = domain $ Short.toShort o
-    origName  d = Short.fromShort $ origDomain d
-    lowerName d = Short.fromShort $ lowerDomain d
+instance IsRepresentation Domain ByteString where
+    fromRepresentation o = domain $ Short.toShort o
+    toRepresentation  d = Short.fromShort $ representation d
 
-instance CaseInsensitiveName Domain String where
-    ciName o = domain $ fromString o
-    origName  d = shortToString $ origDomain d
-    lowerName d = shortToString $ lowerDomain d
+instance IsRepresentation Domain String where
+    fromRepresentation o = domain $ fromString o
+    toRepresentation  d = shortToString $ representation d
 
 -- | append operator using '.'
 --
 -- >>> "www" <.> "example.com"
--- "www.example.com"
+-- "www.example.com."
 -- >>> "com" <.> "."
 -- "com."
 (<.>) :: Domain -> Domain -> Domain
-x <.> "." = x <> "."
+x <.> "." = x
 x <.> y   = x <> "." <> y
 
 infixr 6 <.>
 
 checkDomain :: (ShortByteString -> a) -> Domain -> a
-checkDomain f Domain{..} = f origDomain
+checkDomain f Domain{..} = f representation
 
 modifyDomain :: (ShortByteString -> ShortByteString) -> Domain -> Domain
-modifyDomain f Domain{..} = domain $ f origDomain
+modifyDomain f Domain{..} = domain $ f representation
 
-hasRoot :: Domain -> Bool
-hasRoot d
-  | Short.null o = False
-  | otherwise    = Short.last o == _period
-  where
-    o = origDomain d
+addRoot :: RawDomain -> RawDomain
+addRoot o
+  | Short.null o            = "."
+  | Short.last o == _period = o
+  | otherwise               = o <> "."
 
-addRoot :: Domain -> Domain
-addRoot d
-  | Short.null o            = domain "."
-  | Short.last o == _period = d
-  | otherwise               = domain (o <> ".")
-  where
-   o = origDomain d
+----------------------------------------------------------------
 
-dropRoot :: Domain -> Domain
-dropRoot d
-  | Short.null o            = d
-  | Short.last o == _period = domain $ Short.init o
-  | otherwise               = d
-  where
-   o = origDomain d
```

Comment:
For example, NSEC3 requires a hashing operation on the string of the domain name without the trailing Root '.'.
I would like to have the ability to extract such a representation in this new module.

Reply:
Which should be used for the hash target, the representation format or the wire format?

Reply:
Sorry. I checked again, and there is no problem because it hashes against wire format even in the case of NSEC3.

Reply:
Is `validate` in `dnsext-full-resolver/DNS/Cache/Iterative.hs` necessary?
If not, we can remove `checkDomain`.